## Caltech - Machine Learning
## Australian Apparel Ltd
## Data Wrangling
## Overview
In this Jupyter Notebook, I perform the 'Data Wrangling' used to Load and Inspect the dataset.

<hr/>

### A.  Data Loading and Inspection

In [11]:
# Import necessary libraries
import pandas as pd
import numpy as np

# Load the dataset
def load_data(file_path):
    return pd.read_csv(file_path)

# Inspect the data
def inspect_data(df):
    print("Header 10 rows:")
    print(df.head(10))
    
    print("\nDataFrame Info:")
    print(df.info())
    
    print("\nDataFrame Statistics:")
    print(df.describe())
    
    print("\nMissing Values:")
    print(df.isna().sum())
    
    print("\nIncorrect Values:")
    print(df.notna())

    print(f"\n# of Duplicate rows: {df.duplicated().sum()}")
    
    
   
        



# Load the data
df = load_data('../data/AusApparalSales4thQrt2020.csv')

# Inspect the data
inspect_data(df)



Header 10 rows:
         Date        Time State     Group  Unit  Sales
0  1-Oct-2020     Morning    WA      Kids     8  20000
1  1-Oct-2020     Morning    WA       Men     8  20000
2  1-Oct-2020     Morning    WA     Women     4  10000
3  1-Oct-2020     Morning    WA   Seniors    15  37500
4  1-Oct-2020   Afternoon    WA      Kids     3   7500
5  1-Oct-2020   Afternoon    WA       Men    10  25000
6  1-Oct-2020   Afternoon    WA     Women     3   7500
7  1-Oct-2020   Afternoon    WA   Seniors    11  27500
8  1-Oct-2020     Evening    WA      Kids    15  37500
9  1-Oct-2020     Evening    WA       Men    15  37500

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7560 entries, 0 to 7559
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Date    7560 non-null   object
 1   Time    7560 non-null   object
 2   State   7560 non-null   object
 3   Group   7560 non-null   object
 4   Unit    7560 non-null   int64 
 5  

## Data Inspection Findings:
- No missing values
- 7560 rows across 6 columns [Date, Time, State, Group, Unit, Sales]
- Date range is 10/01/2020 - 12/30/2020
- Sales figures range from $5000 to $162500
- Sales across 7 States
- Sales across 4 Groups:  Kids, Men, Women, and Seniors

<hr/>

### B.  Data Analytics


In [10]:
 print("\nUnique values in Columns:")
for col in df.select_dtypes(include=['object']).columns:
    print(f"\n{col}:")
    print(df[col].unique())
   
# Remove leading or trailing whitespaces from categorical columns
categorical_columns = ['State', 'Group', 'Time']
for col in categorical_columns:
    df[col] = df[col].str.strip()

# Verify there a no leading or trailing whitespaces in String columns
string_columns = df.select_dtypes(include=['object']).columns
for col in string_columns:
    if df[col].str.strip().ne(df[col]).any():
        print(f"Column '{col}' has leading/trailing whitespaces.")
    else:
        print(f"Column '{col}' has no leading/trailing whitespaces.")

# Check for formatting inconsistencies in categorical variables
categorical_columns = ['State', 'Group', 'Time']
for col in categorical_columns:
    print(f"\nUnique values in {col}:")
    print(df[col].unique())     


Unique values in Columns:

Date:
['1-Oct-2020' '2-Oct-2020' '3-Oct-2020' '4-Oct-2020' '5-Oct-2020'
 '6-Oct-2020' '7-Oct-2020' '8-Oct-2020' '9-Oct-2020' '10-Oct-2020'
 '11-Oct-2020' '12-Oct-2020' '13-Oct-2020' '14-Oct-2020' '15-Oct-2020'
 '16-Oct-2020' '17-Oct-2020' '18-Oct-2020' '19-Oct-2020' '20-Oct-2020'
 '21-Oct-2020' '22-Oct-2020' '23-Oct-2020' '24-Oct-2020' '25-Oct-2020'
 '26-Oct-2020' '27-Oct-2020' '28-Oct-2020' '29-Oct-2020' '30-Oct-2020'
 '1-Nov-2020' '2-Nov-2020' '3-Nov-2020' '4-Nov-2020' '5-Nov-2020'
 '6-Nov-2020' '7-Nov-2020' '8-Nov-2020' '9-Nov-2020' '10-Nov-2020'
 '11-Nov-2020' '12-Nov-2020' '13-Nov-2020' '14-Nov-2020' '15-Nov-2020'
 '16-Nov-2020' '17-Nov-2020' '18-Nov-2020' '19-Nov-2020' '20-Nov-2020'
 '21-Nov-2020' '22-Nov-2020' '23-Nov-2020' '24-Nov-2020' '25-Nov-2020'
 '26-Nov-2020' '27-Nov-2020' '28-Nov-2020' '29-Nov-2020' '30-Nov-2020'
 '1-Dec-2020' '2-Dec-2020' '3-Dec-2020' '4-Dec-2020' '5-Dec-2020'
 '6-Dec-2020' '7-Dec-2020' '8-Dec-2020' '9-Dec-2020' '10-Dec-2020'

## Data Analytics findings
- I found there were Leading spaces in Group and State columns, added code to remove spaces.

<hr/>

### C.  Data Wrangling and Normalization

In [14]:
# Identify the numerical columns
numerical_columns = df.select_dtypes(include=[np.number]).columns

# Validate data ranges for numerical columns
for col in numerical_columns:
    print(f"\nStatistics for {col}:")
    print(df[col].describe())
    
# Check for negative values in columns
if col in ['Unit', 'Sales']:
    neg_count = (df[col] < 0).sum()
    if neg_count > 0:
        print(f"Warning: {neg_count} negative values found in {col}")
    else:
        print(f"No negative values found in {col}")
# Check for outliers or incorrect data
print(df[['Sales', 'Unit']].describe())

# Simple normalization for 'Sales' and 'Unit' columns
df['Sales_Normalized'] = (df['Sales'] - df['Sales'].min()) / (df['Sales'].max() - df['Sales'].min())
df['Unit_Normalized'] = (df['Unit'] - df['Unit'].min()) / (df['Unit'].max() - df['Unit'].min())

print("\nFirst few rows with normalized columns:")
print(df[['Sales', 'Sales_Normalized', 'Unit', 'Unit_Normalized']].head())

print("\nSummary of normalized columns:")
print(df[['Sales_Normalized', 'Unit_Normalized']].describe())




Statistics for Unit:
count    7560.000000
mean       18.005423
std        12.901403
min         2.000000
25%         8.000000
50%        14.000000
75%        26.000000
max        65.000000
Name: Unit, dtype: float64

Statistics for Sales:
count      7560.000000
mean      45013.558201
std       32253.506944
min        5000.000000
25%       20000.000000
50%       35000.000000
75%       65000.000000
max      162500.000000
Name: Sales, dtype: float64

Statistics for Sales_Normalized:
count    7560.000000
mean        0.254054
std         0.204784
min         0.000000
25%         0.095238
50%         0.190476
75%         0.380952
max         1.000000
Name: Sales_Normalized, dtype: float64

Statistics for Unit_Normalized:
count    7560.000000
mean        0.254054
std         0.204784
min         0.000000
25%         0.095238
50%         0.190476
75%         0.380952
max         1.000000
Name: Unit_Normalized, dtype: float64
               Sales         Unit
count    7560.000000  7560.000000


## Data Wrangling findings
- Utilized Min/Max for Normalization to scale values from 0 to 1.
- A Mean value of 0.254 or 25.4% of the max values
- A Median value of 0.190 which is less than Mean of 0.254 which indicates a positive (right) skewing of the distribution
- A Standard Deviation of .204 indicates a slight variability in the distribution.

<hr/>

### D.  Data Chunking or Merging

In [24]:
# Group By State
state_grouping = df.groupby('State')

# Display summary  fo each state
print("Summary statistics by State:")
print(state_grouping['Sales'].describe())

# Group By Customer Group
customer_grouping = df.groupby('Group')

# Display summary for each customer grouping
print("\nSummary statistics by Customer Grouping:")
print(customer_grouping['Sales'].describe())

# Group by Month and calculate monthly summaries
monthly_sales = df.groupby('Date').agg({
    'Sales': ['sum', 'mean', 'count'],
    'Unit': ['sum', 'mean']
})

# Rename columns for clarity
monthly_sales.columns = ['Total_Sales', 'Avg_Daily_Sales', 'Num_Transactions', 'Total_Units', 'Avg_Daily_Units']

# Reset index to make 'Month' a column
monthly_sales = monthly_sales.reset_index()

# Calculate Monthly growth rate for Total_Sales
monthly_sales['Sales_Growth_Rate'] = monthly_sales['Total_Sales'].pct_change() * 100

# Display the monthly summaries
print("Monthly Sales Summaries:")
print(monthly_sales)

Summary statistics by State:
        count          mean           std      min      25%      50%  \
State                                                                  
NSW    1080.0  69416.666667  20626.651646  30000.0  52500.0  70000.0   
NT     1080.0  20907.407407   8961.907893   5000.0  15000.0  20000.0   
QLD    1080.0  30942.129630  13344.638002   7500.0  20000.0  30000.0   
SA     1080.0  54497.685185  17460.965183  25000.0  40000.0  52500.0   
TAS    1080.0  21074.074074   9024.684205   5000.0  15000.0  20000.0   
VIC    1080.0  97745.370370  26621.597092  50000.0  77500.0  95000.0   
WA     1080.0  20511.574074   9231.905897   5000.0  12500.0  20000.0   

            75%       max  
State                      
NSW     85000.0  112500.0  
NT      27500.0   37500.0  
QLD     40000.0   62500.0  
SA      67500.0   87500.0  
TAS     27500.0   37500.0  
VIC    112500.0  162500.0  
WA      27500.0   37500.0  

Summary statistics by Customer Grouping:
          count          mea

## Data Chunking and Merging findings
- Usage of the GroupBy function to Analyze the data
- Sales seem to peak in December likely due to Holiday sales